llama-rs/src/lib.rs

```diff
@@ -678,182 +735,132 @@ impl Model {
-                let InferenceParameters {
-                    top_k,
-                    top_p,
-                    repeat_penalty,
-                    temp,
-                    ..
-                } = params;
-
-                let n_vocab = self.hparams.n_vocab;
-
-                let id = self.sample_top_p_top_k(
-                    vocab,
-                    &logits[logits.len() - n_vocab as usize..],
-                    &last_n_tokens,
-                    *repeat_penalty as f64,
-                    *top_k,
-                    *top_p as f64,
-                    *temp as f64,
-                    rng,
-                );
+    pub fn infer_next_token<'v>(
+        &mut self,
+        vocab: &'v Vocabulary,
+        params: &InferenceParameters,
+        rng: &mut impl rand::Rng,
+    ) -> OutputToken<'v> {
+        // First, sample the next token, using the stored last_logits;
+        let next_token = self.sample_top_p_top_k(params, rng);
 
-                last_n_tokens.remove(0);
-                last_n_tokens.push(id);
+        // Update the last_n_tokens list
+        self.last_n_tokens.push_front(next_token);
 
-                // add it to the context
-                embd.push(id);
+        // Then, evaluate the network again to compute the new last_logits
+        self.evaluate(params.n_threads, &[next_token]);
 
-                // decrement remaining sampling budget
-                remaining_tokens -= 1;
-            } else {
-                // if here, it means we are still processing the input prompt
-                while embd_inp.len() > input_consumed {
-                    embd.push(embd_inp[input_consumed]);
-                    last_n_tokens.remove(0);
-                    last_n_tokens.push(embd_inp[input_consumed]);
-                    input_consumed += 1;
-                    if embd.len() > params.n_batch {
-                        break;
-                    }
-                }
-            }
+        // Return the next token
+        if next_token == 2 {
+            OutputToken::EndOfText
+        } else {
+            OutputToken::Token(&vocab.mapping[next_token as usize])
+        }
+    }
 
-            // display text
-            let mut eot = false;
-            for &id in &embd {
-                let output_token = if id == 2 {
-                    eot = true;
-                    OutputToken::EndOfText
-                } else {
-                    OutputToken::Token(&vocab.mapping[id as usize])
-                };
-                callback(output_token);
-            }
+    pub fn inference_with_prompt(
+        &mut self,
+        vocab: &Vocabulary,
+        params: &InferenceParameters,
+        prompt: &str,
+        rng: &mut impl rand::Rng,
+        callback: impl Fn(OutputToken),
+    ) {
+        self.feed_prompt(vocab, params, prompt, |tk| callback(tk));
 
-            if eot {
+        while self.n_past < self.hparams.n_ctx as usize {
```

Comment:
This loop seems a little weird to me. Maybe have a comment explaining what it's doing for people who look at it to figure out how to structure their own loops?

Reply:
The idea is to continue inferring tokens until the model returns an EndOfText, or we run out of space. Whatever happens first.